packages/ethpm-v3/lib/utils.js

```diff
@@ -0,0 +1,301 @@
+  for (let file of files) {
+    const fileContents = JSON.parse(
+      fs.readFileSync(
+        path.join(options.contracts_build_directory, file),
+        "utf8"
+      )
+    );
+    sourcePaths[fileContents.sourcePath] = fileContents.source;
+    detectedArtifacts[fileContents.sourcePath] = fileContents;
+    const allDeploymentNetworkIds = Object.keys(fileContents.networks);
+    allDeploymentNetworkIds.forEach(deploymentNetworkId => {
+      detectedNetworkIdsToBlockchainUris[deploymentNetworkId] = "";
+    });
+  }
+
+  const requiredNetworkIds = Object.keys(detectedNetworkIdsToBlockchainUris);
+  const availableNetworkIds = Object.keys(options.networks).map(n =>
+    options.networks[n].network_id.toString()
+  );
+  for (let networkId of requiredNetworkIds) {
+    if (!availableNetworkIds.includes(networkId)) {
+      throw new TruffleError(
+        `Detected a deployment for network id: ${networkId}, but missing a provider for this network. Please make sure there is an available provider for this network id in your truffle-config.`
+      );
+    }
+  }
+
+  // generate blockchainURI for each detected networkId
+  for (let networkId of requiredNetworkIds) {
+    let targetProvider;
+    for (let provider of Object.keys(options.networks)) {
+      if (options.networks[provider].network_id == networkId) {
+        // handle ganache provider from tests
+        if (
+          options.networks[provider].provider.constructor.name == "Function"
+        ) {
+          targetProvider = options.networks[provider].provider();
+        } else {
+          targetProvider = options.networks[provider].provider;
+        }
+      }
+    }
+    const blockchainUri = await getBlockchainUriForProvider(targetProvider);
+    detectedNetworkIdsToBlockchainUris[networkId] = blockchainUri;
+  }
+
+  // replace network ids with blockchain uri in detected artifacts
+  for (let artifactPath of Object.keys(detectedArtifacts)) {
+    var artifactDeployments = detectedArtifacts[artifactPath].networks;
+    var updatedDeployments = {};
+    Object.keys(artifactDeployments).forEach(id => {
+      var matchingUri = detectedNetworkIdsToBlockchainUris[id];
+      updatedDeployments[matchingUri] = artifactDeployments[id];
+    });
+    detectedArtifacts[artifactPath].networks = updatedDeployments;
+  }
+
+  const normalizedArtifacts = Object.keys(detectedArtifacts).map(
+    key => detectedArtifacts[key]
+  );
+  const parsedArtifacts = parseTruffleArtifacts(normalizedArtifacts);
+  const sources = await resolveSources(
+    sourcePaths,
+    options.contracts_directory,
+    ethpm
+  );
+  return {
+    resolvedSources: sources,
+    resolvedContractTypes: parsedArtifacts.contractTypes,
+    resolvedDeployments: parsedArtifacts.deployments,
+    resolvedCompilers: parsedArtifacts.compilers
+  };
+}
+
+async function resolveEnsName(address, provider, options) {
+  let resolvedAddress;
+  const w3 = new Web3(provider);
```

Comment:
Convention in Truffle is to use `web3` as the variable name for the instance of `Web3`.